### DISCUSSION TITLE
Content Repository for the CMS!?

### DISCUSSION THREAD

**Discussion by SuperFlySly** [pimcore/pimcore#17354] [Ideas]
I am enthusiastic about Pimcore and I think Pimcore is one of the best open source projects. However, in my opinion, what is urgently needed is a content repository for the CMS area. This feature would make Pimcore even more powerful and flexible than it already is. It would also solve a lot of problems, especially when it comes to updates or reusing content for different areas.

CMS systems that already stand out by providing a content repository are for example the Neos-CMS or the Sulu-CMS.

A content repository would make Pimcore a real game changer. 

What do you think?

**Comment by kingjia90** [comment]
Thank you for your feedback!

Out of personal curiosity, what is a Content Repository? How would it differ from the [Assets](https://pimcore.com/docs/platform/Pimcore/Assets/) + [Direct Edit](https://pimcore.com/docs/platform/Direct_Edit/)? Isn't the same thing at some degrees?

**Comment by SuperFlySly** [comment]
Hi, you can find out what a content repo is here 

https://en.m.wikipedia.org/wiki/Content_repository

and in my opinion the best and most advanced content repo implementation is Neos Content Repository: https://docs.neos.io/guide/content-repository/data-structure

especially because of the Concept of „Content Dimensions“.

**Reply by dpfaffenbauer** [reply]
You got that in Pimcore with Documents...

**Reply by SuperFlySly** [reply]
No, you don’t — not at all!

**Reply by dpfaffenbauer** [reply]
why not? you can define a data structure with bricks, and you can use the backend to fill the content

**Reply by SuperFlySly** [reply]
I used to work at elements in Salzburg and updating and migrating content is very painful and time consuming. Apart from that, with the current solution you don't have anywhere near the possibilities that you have with a real content repository.

...so it would be just as justified to store all other objects serialized in the database!

**Reply by dpfaffenbauer** [reply]
ok. whats the difference then?

**Reply by SuperFlySly** [reply]
A content repository provides a structured, centralized solution for managing and storing content. Here are some advantages compared to serialized storage of unstructured data:

1. **Structured Data Storage**:
   - **Content Repository**: Stores data in an organized structure, often in the form of hierarchies or nodes. This makes it easier to find and manage content.
   - **Serialized Storage**: Unstructured data is often stored as large data blocks, making organization and access more challenging.

2. **Query and Search Capabilities**:
   - **Content Repository**: Offers advanced search and query capabilities, often through standardized query languages like SQL or JCR-SQL2.
   - **Serialized Storage**: Often requires additional tools or manual processes to search and retrieve data.

3. **Versioning and History**:
   - **Content Repository**: Supports versioning and tracking of changes, allowing easy retrieval of previous content versions.
   - **Serialized Storage**: Implementing versioning is more complex and less efficient.

4. **Access Control and Security**:
   - **Content Repository**: Provides integrated mechanisms for managing permissions and access controls, enhancing security.
   - **Serialized Storage**: Security mechanisms often need to be implemented separately, leading to potential inconsistencies.

5. **Scalability and Performance**:
   - **Content Repository**: Optimized for high scalability and performance for concurrent read and write operations.
   - **Serialized Storage**: Can quickly reach performance limits with large data volumes and high user activity.

6. **Transaction Management**:
   - **Content Repository**: Supports transaction management to ensure data consistency and integrity.
   - **Serialized Storage**: Transaction management often needs to be implemented separately, which can be complex and error-prone.

7. **Integration and Interoperability**:
   - **Content Repository**: Offers standardized APIs and integration capabilities with other systems and applications.
   - **Serialized Storage**: Integration with other systems is often more labor-intensive and requires additional customization.

Therefore, a content repository is often more efficient and user-friendly than serialized storage of unstructured data, especially when managing large amounts of content.

**Reply by fashxp** [reply]
data objects?

**Reply by SuperFlySly** [reply]
Well, if the solution is so simple, why is the data serialized in the CMS area?

**Reply by fashxp** [reply]
Different usecase...

**Reply by SuperFlySly** [reply]
Exactly! And considering the fact that Pimcore is very often used as a CMS, a content repository would make a lot of sense!

**Reply by SuperFlySly** [reply]
...because in this case everything could really be stored consistently...

**Reply by fashxp** [reply]
But why not use data objects when you need the things described above, and use documents when you need more fexibility and freedom? And combine the things where needed...

**Reply by twin-elements** [reply]
@SuperFlySly Do I understand correctly that your main concern is searching through documents?

**Reply by dpfaffenbauer** [reply]
@SuperFlySly Neos Content Repository is also unstructured, quote:

> The content in Neos is not stored inside tables of a relational database, but inside a tree-based structure: the so-called Neos Content Repository.
> To a certain extent, it is comparable to files in a file-system: They are also structured as a tree, and are identified uniquely by the complete path towards the file.

Similar to Pimcore Documents, you have the FullPath, and then the path to the editable.

**Reply by SuperFlySly** [reply]
Ok guys, I would like to use content inheritance and would love it if I could use Pimcore out-of-box as a headless CMS, until then I have to fall back on the Sulu CMS.

**Reply by dpfaffenbauer** [reply]
Content Inheritance is supported. Headless as well. Headless is just JSON rendering instead of HTML ;). Install dachcom toolbox that has headless support out of the box.
